File: Common/JwtModules/CsJwt/JwtReader.cs
using System;
using System.Collections.Generic;
using System.Security.Cryptography;
using System.Text;
using System.Text.Json;
using System.Text.RegularExpressions;

namespace JwtReader;

public static class JwtHelpers
{
    private static readonly Regex UsernameRegex = new(@"^[a-zA-Z0-9_]{3,30}$", RegexOptions.Compiled);

    public static string ExtractBearerToken(string authorizationHeader)
    {
        if (string.IsNullOrWhiteSpace(authorizationHeader) ||
            !authorizationHeader.StartsWith("Bearer ", StringComparison.OrdinalIgnoreCase))
        {
            throw new ArgumentException("Authorization header must start with Bearer ");
        }

        var token = authorizationHeader["Bearer ".Length..].Trim();
        if (string.IsNullOrWhiteSpace(token))
            throw new ArgumentException("Empty bearer token");

        return token;
    }

    public static JwtToken ReadAndValidate(string authorizationHeader, string secret, long? nowUnix = null)
    {
        var token = ExtractBearerToken(authorizationHeader);
        return ParseAndValidate(token, secret, nowUnix ?? DateTimeOffset.UtcNow.ToUnixTimeSeconds());
    }

    public static JwtToken ParseAndValidate(string jwt, string secret, long nowUnix)
    {
        if (string.IsNullOrWhiteSpace(secret))
            throw new ArgumentException("JWT secret is empty");

        var parts = jwt.Split('.');
        if (parts.Length != 3)
            throw new FormatException("Invalid JWT format");

        var headerJson = Encoding.UTF8.GetString(Base64UrlDecode(parts[0]));
        var payloadJson = Encoding.UTF8.GetString(Base64UrlDecode(parts[1]));

        var header = JsonSerializer.Deserialize<JwtHeader>(headerJson)
                     ?? throw new FormatException("Invalid JWT header");
        if (!string.Equals(header.alg, "HS256", StringComparison.Ordinal))
            throw new NotSupportedException("Unsupported algorithm; expected HS256");

        if (!VerifyHs256($"{parts[0]}.{parts[1]}", parts[2], secret))
            throw new CryptographicException("Invalid JWT signature");

        var claims = JsonSerializer.Deserialize<JwtClaims>(payloadJson)
                     ?? throw new FormatException("Invalid JWT payload");

        ValidateClaims(claims, nowUnix);

        return new JwtToken
        {
            Raw = jwt,
            Header = header,
            Claims = claims,
            Subject = claims.sub
        };
    }

    private static bool VerifyHs256(string signingInput, string signatureB64Url, string secret)
    {
        using var hmac = new HMACSHA256(Encoding.UTF8.GetBytes(secret));
        var expected = hmac.ComputeHash(Encoding.UTF8.GetBytes(signingInput));
        var got = Base64UrlDecode(signatureB64Url);
        return CryptographicOperations.FixedTimeEquals(expected, got);
    }

    private static void ValidateClaims(JwtClaims c, long nowUnix)
    {
        if (string.IsNullOrWhiteSpace(c.user_id) ||
            string.IsNullOrWhiteSpace(c.username) ||
            string.IsNullOrWhiteSpace(c.email) ||
            string.IsNullOrWhiteSpace(c.role) ||
            string.IsNullOrWhiteSpace(c.sub))
            throw new InvalidOperationException("Missing required claim");

        if (c.iat == 0 || c.exp == 0 || c.nbf == 0)
            throw new InvalidOperationException("Missing required timestamp claim");

        if (c.role != "user" && c.role != "admin")
            throw new InvalidOperationException("Invalid role");

        if (!UsernameRegex.IsMatch(c.username))
            throw new InvalidOperationException("Invalid username");

        if (c.exp < nowUnix)
            throw new InvalidOperationException("Token expired");

        if (c.nbf > nowUnix)
            throw new InvalidOperationException("Token not yet valid");

        if (c.user_id != c.sub)
            throw new InvalidOperationException("user_id must match sub");
    }

    private static byte[] Base64UrlDecode(string input)
    {
        string s = input.Replace('-', '+').Replace('_', '/');
        switch (s.Length % 4)
        {
            case 2: s += "=="; break;
            case 3: s += "="; break;
            case 1: throw new FormatException("Invalid base64url string");
        }
        return Convert.FromBase64String(s);
    }
}

public sealed class JwtHeader
{
    public string alg { get; set; } = "";
    public string typ { get; set; } = "";
}

public sealed class JwtClaims
{
    public string user_id { get; set; } = "";
    public string username { get; set; } = "";
    public string email { get; set; } = "";
    public string role { get; set; } = "";

    public string sub { get; set; } = "";
    public long iat { get; set; }
    public long exp { get; set; }
    public long nbf { get; set; }
    public string? jti { get; set; }
}

public sealed class JwtToken
{
    public string Raw { get; set; } = "";
    public JwtHeader Header { get; set; } = new();
    public JwtClaims Claims { get; set; } = new();
    public string Subject { get; set; } = "";
}

File: Common/JwtModules/JavaJwt/JwtReader.java
package jwtreader;

import com.fasterxml.jackson.databind.DeserializationFeature;
import com.fasterxml.jackson.databind.ObjectMapper;

import javax.crypto.Mac;
import javax.crypto.spec.SecretKeySpec;
import java.nio.charset.StandardCharsets;
import java.time.Instant;
import java.util.Base64;
import java.util.regex.Pattern;

public final class JwtReader {
    private static final Pattern USERNAME_PATTERN = Pattern.compile("^[a-zA-Z0-9_]{3,30}$");
    private static final ObjectMapper MAPPER = new ObjectMapper()
            .configure(DeserializationFeature.FAIL_ON_UNKNOWN_PROPERTIES, false);

    private JwtReader() {}

    public static String extractBearerToken(String authorizationHeader) {
        if (authorizationHeader == null || !authorizationHeader.startsWith("Bearer ")) {
            throw new IllegalArgumentException("Authorization header must start with Bearer ");
        }
        String token = authorizationHeader.substring("Bearer ".length()).trim();
        if (token.isEmpty()) {
            throw new IllegalArgumentException("Empty bearer token");
        }
        return token;
    }

    public static JwtToken readAndValidate(String authorizationHeader, String secret) {
        return parseAndValidate(extractBearerToken(authorizationHeader), secret, Instant.now().getEpochSecond());
    }

    public static JwtToken parseAndValidate(String jwt, String secret, long nowUnix) {
        if (secret == null || secret.isEmpty()) {
            throw new IllegalArgumentException("JWT secret is empty");
        }

        String[] parts = jwt.split("\\.");
        if (parts.length != 3) {
            throw new IllegalArgumentException("Invalid JWT format");
        }

        try {
            String headerJson = new String(base64UrlDecode(parts[0]), StandardCharsets.UTF_8);
            String payloadJson = new String(base64UrlDecode(parts[1]), StandardCharsets.UTF_8);

            JwtHeader header = MAPPER.readValue(headerJson, JwtHeader.class);
            if (!"HS256".equals(header.alg)) {
                throw new IllegalArgumentException("Unsupported algorithm; expected HS256");
            }

            if (!verifyHs256(parts[0] + "." + parts[1], parts[2], secret)) {
                throw new SecurityException("Invalid JWT signature");
            }

            JwtClaims claims = MAPPER.readValue(payloadJson, JwtClaims.class);
            validateClaims(claims, nowUnix);

            return new JwtToken(jwt, header, claims);
        } catch (Exception e) {
            if (e instanceof RuntimeException re) throw re;
            throw new RuntimeException("Failed to parse JWT: " + e.getMessage(), e);
        }
    }

    private static boolean verifyHs256(String signingInput, String signatureB64Url, String secret) throws Exception {
        Mac mac = Mac.getInstance("HmacSHA256");
        mac.init(new SecretKeySpec(secret.getBytes(StandardCharsets.UTF_8), "HmacSHA256"));
        byte[] expected = mac.doFinal(signingInput.getBytes(StandardCharsets.UTF_8));
        byte[] got = base64UrlDecode(signatureB64Url);
        return java.security.MessageDigest.isEqual(expected, got);
    }

    private static void validateClaims(JwtClaims c, long nowUnix) {
        if (isBlank(c.user_id) || isBlank(c.username) || isBlank(c.email) || isBlank(c.role) || isBlank(c.sub)) {
            throw new IllegalArgumentException("Missing required claim");
        }
        if (c.iat == 0 || c.exp == 0 || c.nbf == 0) {
            throw new IllegalArgumentException("Missing required timestamp claim");
        }
        if (!"user".equals(c.role) && !"admin".equals(c.role)) {
            throw new IllegalArgumentException("Invalid role");
        }
        if (!USERNAME_PATTERN.matcher(c.username).matches()) {
            throw new IllegalArgumentException("Invalid username");
        }
        if (c.exp < nowUnix) {
            throw new IllegalArgumentException("Token expired");
        }
        if (c.nbf > nowUnix) {
            throw new IllegalArgumentException("Token not yet valid");
        }
        try {
            if (Long.parseLong(c.user_id) != Long.parseLong(c.sub)) {
                throw new IllegalArgumentException("user_id must match sub");
            }
        } catch (NumberFormatException e) {
            throw new IllegalArgumentException("user_id and sub must be numeric strings");
        }
    }

    private static boolean isBlank(String s) {
        return s == null || s.trim().isEmpty();
    }

    private static byte[] base64UrlDecode(String s) {
        return Base64.getUrlDecoder().decode(s);
    }

    public static final class JwtHeader {
        public String alg = "";
        public String typ = "";
    }

    public static final class JwtClaims {
        public String user_id = "";
        public String username = "";
        public String email = "";
        public String role = "";
        public String sub;
        public long iat;
        public long exp;
        public long nbf;
        public String jti;
    }

    public static final class JwtToken {
        public final String raw;
        public final JwtHeader header;
        public final JwtClaims claims;

        public JwtToken(String raw, JwtHeader header, JwtClaims claims) {
            this.raw = raw;
            this.header = header;
            this.claims = claims;
        }
    }
}
